classes/order/Order.php

```diff
@@ -1852,7 +1852,7 @@ public function addOrderPayment($amount_paid, $payment_method = null, $payment_t
         if ($order_payment->id_currency == $this->id_currency) {
             $this->total_paid_real += $order_payment->amount;
         } else {
-            $this->total_paid_real += Tools::ps_round(Tools::convertPrice($order_payment->amount, $order_payment->id_currency, false), 2);
+            $this->total_paid_real += Tools::ps_round(Tools::convertPrice($order_payment->amount, $this->id_currency, false), 2);
```

Comment:
Ok so after digging into the code I think I finally understood your modif which seems correct

The confusion comes from the fact that OrderPayment's currency matches the `$currency` parameter (this is correct), but `total_paid_real` must be expressed as the Order's currency, hence the use of `$this->id_currency` The gymnastic is not so simple at first I think a comment describing this behavoiur would be a good thing :)

Reply:
I just thought of an additional use case, what happens when the Order AND the Order payment are different from the default currency I think the conversion rate won't match

The amount should first be converted back to the default currency value, and then converted into the third currency value

Reply:
Unless the Order->total_paid_real value is stored as the default value? (as the third parameter `$to_currency` suggests
Then it should be correct, but out of safety we should have a behat test to valide this:
- create an order with currency different from the default one
- pay with a currency different from the default one
- check that everything is correctly converted into the default currency

Reply:
But since Order has a currency_id as well, I would tend to think that its value is also expressed as this currency value, not the default one

Reply:
```suggestion
            // Convert payment value as default currency
            $paymentDefaultValue = Tools::ps_round(Tools::convertPrice($order_payment->amount, $order_payment->id_currency, false), 2);
            // Then convert into the order currency's value with initially saved conversion rate
            $paymentOrderValue = $paymentDefaultValue * $this->conversion_rate;
            $this->total_paid_real += $paymentOrderValue;
```

I also noticed that conversion rate is saved when the order is created, I'm not sure if we should use the rate at the moment the order was created or when it was paid..

Reply:
> I also noticed that conversion rate is saved when the order is created, I'm not sure if we should use the rate at the moment the order was created or when it was paid..

the conversion rate is the rate between order's currency and the default currency. But the default currency can be changed, so we cannot use this conversion rate
Validated with @MatShir

Reply:
> Unless the Order->total_paid_real value is stored as the default value? (as the third parameter `$to_currency` suggests
> Then it should be correct, but out of safety we should have a behat test to valide this:
> 
> * create an order with currency different from the default one
> * pay with a currency different from the default one
> * check that everything is correctly converted into the default currency

I'm creating the behat test but the conversion made is not from "payment currency" to "default currency", but from "payment currency" to "order currency". So whatever the default currency is, I think it will be ok

But you're right to ask "what about the conversion_rate ?"

Reply:
@jolelievre You was absolutely right. We need to convert to default currency before converting to the order one, because of the logic of convertPrice and how the conversion rate is set to the currencies

And that make me understand why you said that changing a shop default currency is not a good idea